app/models/order_import.rb

```diff
@@ -1,142 +1,101 @@
+  # end
+  #
+  # Be sure to honor #send_receipts
+  #
+
+  def process_upload!
+    init_error_report
 
     if self.fail_on_error?
-      result = handle_save_nothing_on_error(upload_file_path, result)
+      handle_save_nothing_on_error
     else
-      result = handle_save_clean_orders(upload_file_path, result)
+      handle_save_clean_orders
     end
 
-    # write error_report to error_file if failed
-    if result.failed?
-      self.error_file = StoredFile.new(
-        :file       => StringIO.new(self.error_report),
-        :file_type  => "import_error",
-        :name => "error_report.csv",
-        :created_by => creator.id
-      )
-
-      self.error_file.file.instance_write(:file_name, "error_report.csv")
-      self.error_file.save!
-    end
+    store_error_report if result.failed?
     self.save!
-
-    return result
   end
 
-  def handle_save_nothing_on_error(upload_file_path, result)
-    # loop over non-header rows
-    Order.transaction do
-      CSV.open(upload_file_path, :headers => true).each do |row|
-        row_errors = errors_for(row)
-
-        # write to error_report in case an error occurs
-        row[ERRORS_HEADER] = row_errors.join(", ")
-        self.error_report += row.to_csv
-
-        if row_errors.length > 0
-          result.failures += 1
-        else
-          result.successes += 1
-        end
-      end
-
-      raise ActiveRecord::Rollback if result.failed?
-    end
-
-    unless result.failed?
-      if self.send_receipts?
-        # send notifications
-        order_ids = @order_id_cache_by_order_key.values()
-        orders    = Order.where(:id => order_ids)
-        orders.each do |order|
-          Notifier.order_receipt(:user => order.user, :order => order).deliver
-        end
-      end
+  def result
+    @result ||= Result.new
+  end
 
-      # we didn't fail, throw away the error report/file
-      self.error_file = nil
-      self.error_report = nil
-    end
+  def error_file_content
+    error_file.try(:read)
+  end
 
-    return result
+  private
+
+  def create_order_from_imported_row!(row_importer)
+    Order.create!(
+      facility: facility,
+      account: row_importer.account,
+      user: row_importer.user,
+      created_by_user: creator,
+      ordered_at: row_importer.order_date,
+      order_import_id: self.id,
+    )
   end
 
-  def handle_save_clean_orders(upload_file_path, result)
-    # build rows_by_order_key ( order_key => [row+] )
-    rows_by_order_key = Hash.new{|h, k| h[k] = []}
-    CSV.open(upload_file_path, :headers => true).each do |row|
-      order_key = [row[USER_HEADER], row[CHART_STRING_HEADER], row[ORDER_DATE_HEADER]]
-      rows_by_order_key[order_key] << row
-    end
+  def discard_error_report
+    self.error_file = nil
+    self.error_report = nil
+  end
 
+  def error_mode?
+    @in_error_mode
+  end
 
-    # loop over the order keys
+  def handle_save_clean_orders # TODO refactor and rename
```

Comment:
Is there anything we can do to make `handle_save_clean_orders` and `handle_save_nothing_error` more inline with each other? One is doing the CSV reading, while the other is delegating out to another method. Definitely makes it hard to follow.

If that's what you meant by "refactor and rename" then we can hold off for a little bit.